F not via

Superdroidz opened the pull request:
The range query doesn't have any tests currently in general. Will add some to test not vias but not sure what the original correct output is suppose to be.

Superdroidz committed:
Add not-via argument to scan method to stop scanning if a stop equals the not via stop.
Add not-via argument to first-arrival endpoint
Added launch configuration for running Mocha Test
Made all queries other than DepartAfterQuery take an empty list.
Add three not via tests
Corrected tests using plan

Superdroidz committed:
Merge branch 'dev' into f-not-via

Superdroidz committed:
Added notVias to scanRange
Made optional with a default in plan methods in Queries.
Added optional retrieval of notVias in endpoint methods.

Superdroidz committed:
Remove unnecessary empty lists for not vias.

iblocks-Xander commented on src/raptor/RaptorAlgorithm.ts:
```diff
@@ -23,7 +23,8 @@ export class RaptorAlgorithm {
     dow: DayOfWeek,
-    time: Time
+    time: Time,
+    notVias: Stop[]
```
I would have made the default value here, rather than further up in the API personally.

Superdroidz replied:
So would you rather have the default here and it required the higher ups or it have a default value all the way up?

iblocks-Xander replied:
I would have it default here, then it essentially becomes optional all the way up. It means that the other APIs wouldn't have changed in this instance.
But I'm not too hung up on changing this, given that you have changed all the API sigs anyway.

Superdroidz replied:
8c8b9d7 Ive added it to be optional here and removed a case of an empty []. The others have renamed as optional defaults where they are needed i beleive

iblocks-Xander replied:
Looks good.

iblocks-Xander commented on test/unit/raptor/RaptorDepartAfterQuery.spec.ts:
```diff
@@ -29,6 +29,98 @@ describe("RaptorDepartAfterQuery", () => {
     ]);
   });
 
+  it("finds route not through via simple", () => {
```
I'd make this test name a bit more descriptive such as: "should find a slower journey to not go via B".

Superdroidz replied:
8c8b9d7 updated name

iblocks-Xander commented on test/unit/raptor/RaptorDepartAfterQuery.spec.ts:
```diff
@@ -29,6 +29,98 @@ describe("RaptorDepartAfterQuery", () => {
+          ]};
+
+      const raptor = RaptorQueryFactory.createDepartAfterQuery(trips, transfers, {}, calendars, journeyFactory);
+      const result = raptor.plan("A", "D", new Date("2018-10-16"), 900, ["B"]);
```
Missing test cases for multiple not via stations.

Superdroidz replied:
8c8b9d7 added multiple not via test

iblocks-Xander commented on src/runner.ts:
```diff
@@ -67,9 +70,12 @@ loadingNetwork.then(([trips, transfers, interchange, calendars]) => {
-    console.log(orig, dest, searchDate, startDate, endDate);
 
-    const journeys = detailedQuery.plan(orig, dest, searchDate, startDate, endDate);
+    const notVias = req.query.notVia ? req.query.notVia : [];
```
We should sanity check the origin and destination against the notVia list, I have a feeling that this will break the algorithm further down.

iblocks-Xander replied:
Also, it means we can reply with a sane error message instead of perhaps an empty result!

Superdroidz replied:
8c8b9d7 ive added checks to this in runner.ts but i was wondering if there should be additional checks in the raptorAlgortithm it self since any future uses wouldnt guard against this. Do you think its worth it throwing an Error in both scan methods?

iblocks-Xander replied:
I have mixed feelings about throwing an error in the scans. 
It's nice because you can keep that logic in one place. On the other hand, there's no way to signal that this will throw an error in the code nicely. 

I'm sure it's not idiomatic JS/TS to expect to handle errors like that. Usually, there's an error callback that is passed in, and the function usually calls the error function if it errors. 

In this case, I'd also argue that we shouldn't throw an error, we should return something more helpful. We should return a `400` response at the very least!

Superdroidz replied:
7fb6b84 corrected to send a 400 response when it finds the origin or destination in the not via list.
 Body of message will just the message. E.g: Origin: BTN in not via list: [BTN]

Logs:
{
  orig: 'BTN',
  dest: 'LBG',
  startDate: '2019-04-29T10:00:00',
  notVia: '[BTN]'
}
Origin: BTN in not via list: [BTN]
{
  orig: 'BTN',
  dest: 'LBG',
  startDate: '2019-04-29T10:00:00',
  notVia: '[LBG]'
}
Destination: LBG in not via list: [LBG]

iblocks-Xander replied:
Cool.

iblocks-Xander commented on test/unit/raptor/RaptorDepartAfterQuery.spec.ts:
```diff
@@ -29,6 +29,98 @@ describe("RaptorDepartAfterQuery", () => {
+  it("finds correct not via ignoring transfers", () => {
+      const trips = [
+          t(st("A", null, 1000),
+              st("B", 1025, 1028),
```
Is there a way of easily representing a train that goes through a station, but doesn't stop at that station?

Superdroidz replied:
You can create a StopTime that has a false pickUp and dropOff.

What test did you have in mind? I think you still should not be allowed to take the trip that would go through the station regardless if you can or cant get on or off of it.

iblocks-Xander replied:
I agree that a not-via should prevent you from going through that station, even with no passenger activity.

I think create a new function in the utils class like `via`, similar to `st`, have it generate stops with optional pick ups or drop offs. Then add a test making sure that it finds the correct journeys.

Superdroidz replied:
7fb6b84 added new via util function and test.

iblocks-Xander replied:
Cool, looks good!

iblocks-Xander reviewed (commented):
First pass review

Superdroidz committed:
Merge branch 'dev' into f-not-via

Superdroidz committed:
Made the notVias have a default in the scan methods.
Removed the empty list from the use of the scan method in transfer-pattern
Validate notVias that they do not include the origin or destination
Rename tests so they are more descriptive

iblocks-Xander commented:
A few more comments.

Superdroidz committed:
Send a bad request and log the issue if the origin or destination exisits in the not vias.
Added test for testing it correctly doesnt route through a station with no pickups or drop offs.

Superdroidz committed:
Merge branch 'dev' into f-not-via
